src/clojure/nrepl/cmdline.clj

```diff
@@ -239,59 +244,193 @@
+  (exit 0))
+
+(defn get-transport-option
+  "Takes a map of nREPL CLI options.
+  Returns either a default transport or the
+  value of :transport."
+  [options]
+  (or (some->> options
+               (:transport)
+               (require-and-resolve :transport))
+      #'transport/bencode))
+
+(defn get-handler-option
+  "Takes a map of nREPL CLI options and list of middleware.
+  Returns a request handler function.
+  If some handler was explicitly passed we'll use it, otherwise we'll build
+  one from whatever was passed via --middleware"
+  [options middleware]
+  (or (some->> options
+               (:handler)
+               (require-and-resolve :handler))
+      (build-handler middleware)))
+
+(defn get-ack-port
+  "Takes a map of nREPL CLI options.
+  Returns integer ack port or nil."
+  [options]
+  (some-> options
+          (:ack)
+          (->int)))
+
+(defn get-greeting-option
+  "Takes a map of nREPL CLI options and the selected transport for the server.
+  Returns a greeting function or nil."
+  [options transport]
+  (when (= transport #'transport/tty)
+    #'transport/tty-greeting))
+
+(defn get-conn-options
+  "Takes map of nREPL CLI options
+  Returns map of processed options used to connect or start a nREPL server."
+  [options]
+  {:port (->int (:port options))
+   :host (:host options)
+   :transport (get-transport-option options)})
+
+(defn get-server-options
+  "Takes a map of nREPL CLI options
+  Returns map of processed options to start an nREPL server."
+  [options]
+  (let [middleware (sanitize-middleware-option (:middleware options))
+        {:keys [host port transport]} (get-conn-options options)]
+    (merge options
+           {:host host
+            :port port
+            :transport transport
+            :bind (:bind options)
+            :middleware middleware
+            :handler (get-handler-option options middleware)
+            :greeting (get-greeting-option options transport)
+            :ack-port (get-ack-port options)})))
+
+(defn connect-command
+  "Connects to a running nREPL server and runs a REPL. Exits program when REPL
+  is closed.
+  Takes a map of nREPL CLI options."
+  [options]
+  (let [{:keys [host port transport]} (get-conn-options options)]
+    (run-repl host port {:transport transport})
+    (exit 0)))
+
+(defn ack-server
+  "Acknowledge the port of this server to another nREPL server running on
+  :ack port.
+  Takes nREPL server map and processed CLI options map.
+  Prints a message describing the acknowledgement between servers.
+  Returns nil."
+  [server options]
+  (when-let [ack-port (get-ack-port options)]
+    (let [port (:port server)
+          transport (:transport options)]
+      (binding [*out* *err*]
+        (println (format "ack'ing my port %d to other server running on port %d"
+                         port ack-port)
+                 (:status (send-ack port ack-port transport)))))))
+
+(defn print-connection-header
```

Comment:
Probably here I'd also decouple the message generation from the message printing.